### PR TITLE
Fix to ignore empty captures from params

### LINKED ISSUE #1388: Before filter adds empty captures to params
Using a before filter with no pattern adds an empty `captures` attribute to `params`. This is unexpected behaviour when your application doesn't contain any regexp patterns with captures, and can cause problems for applications which are mapping `params` directly to another object.

A minimal case: 

```ruby
require 'sinatra'

before do
end

get '/' do
  params.to_s
end
```

```sh
$ curl -i localhost:9292
HTTP/1.1 200 OK 
Content-Type: text/html;charset=utf-8
Content-Length: 16
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
Vary: Accept-Encoding
Server: WEBrick/1.3.1 (Ruby/2.3.3/2016-11-21)
Date: Thu, 15 Feb 2018 10:20:14 GMT
Connection: Keep-Alive

{"captures"=>[]}
```
 I suspect this might also be the cause of issue https://github.com/sinatra/sinatra/issues/1380. 

I think it's happening due to this [line of code](https://github.com/sinatra/sinatra/blob/62fb8ad8c6a42a2e1b755a9cbc419b78ea5f10ea/lib/sinatra/base.rb#L1346), which defaults the before filter pattern to the regexp `/.*/`:

```ruby
def before(path = /.*/, **options, &block)
        add_filter(:before, path, options, &block)
end
```

### PR DESCRIPTION
**Pull request by iguchi1124**
I'm trying to fix #1388 

`params` should be empty hash if requests doesn't matches any regex of routings. But It was returned `{"captures" => []}`. so I fixed to return empty hash object.

```rb
before do
  # noop
end

get "/" do
  params.to_s # returns `{}`
end
```

### CODE REVIEW

File: lib/sinatra/base.rb [hunk 1]
```diff
@@ -1343,19 +1343,19 @@ def mime_types(type)
...
       # context as route handlers and may access/modify the request and
       # response.
-      def after(path = /.*/, **options, &block)
+      def after(path = /(.*)/, **options, &block)
         add_filter(:after, path, options, &block)
       end
 
       # add a filter
-      def add_filter(type, path = /.*/, **options, &block)
+      def add_filter(type, path, **options, &block)
```
Comment (iguchi1124) [inline lib/sinatra/base.rb]: I think this method's argument of `path` doesn't need default value

**Comment by namusyaka** [comment]
@iguchi1124 Thank you for the patch, I'll take care of it soon.

File: lib/sinatra/base.rb [hunk 2]
```diff
@@ -1343,19 +1343,19 @@ def mime_types(type)
...
-      def before(path = /.*/, **options, &block)
+      def before(path = /(.*)/, **options, &block)
         add_filter(:before, path, options, &block)
       end
 
       # Define an after filter; runs after all requests within the same
       # context as route handlers and may access/modify the request and
       # response.
-      def after(path = /.*/, **options, &block)
+      def after(path = /(.*)/, **options, &block)
```
Comment (namusyaka) [inline lib/sinatra/base.rb]: You shouldn't change this regexp.

File: lib/sinatra/base.rb (hunk 1 above)
Comment (namusyaka) [reply lib/sinatra/base.rb]: Potentially breaking compatibility, I don't think there is a reason for breaking it.

**Comment by namusyaka** [comment]
@iguchi1124 If you're still planning on addressing the problem, please consider the following points:

- Needs a test.
- I guess the patch would be something like:

```diff
diff --git a/lib/sinatra/base.rb b/lib/sinatra/base.rb
index b2dff314..5d1a09f3 100644
--- a/lib/sinatra/base.rb
+++ b/lib/sinatra/base.rb
@@ -1029,7 +1029,7 @@ module Sinatra
       if regexp_exists
         captures           = pattern.match(route).captures
         values            += captures
-        @params[:captures] = captures
+        @params[:captures] = captures if captures && !captures.empty?
       else
         values += params.values.flatten
       end
```

**Comment by iguchi1124** [comment]
Thanks! Updated pull request description and code changes!

File: lib/sinatra/base.rb [hunk 3]
```diff
@@ -1029,7 +1029,7 @@ def process_route(pattern, conditions, block = nil, values = [])
       if regexp_exists
         captures           = pattern.match(route).captures
         values            += captures
-        @params[:captures] = captures
+        @params[:captures] = captures if !captures.nil? && !captures.empty?
```
Comment (namusyaka) [inline lib/sinatra/base.rb]: If you hope that the captures checking is using `Object#nil?`, you could simplify the condition by using De Morgan's laws.
`unless captures.nil? || captures.empty?`

Comment (iguchi1124) [reply lib/sinatra/base.rb]: refactored and fixup in this commit https://github.com/sinatra/sinatra/pull/1390/commits/51fe0fcb00cf8f21c524370a3b7c7a4a4c7d067a

**Comment by namusyaka** [comment]
Great!

**Pull request merged**